docs/advanced/authentication.md

```diff
@@ -0,0 +1,328 @@
+
+```js
+const getAuth = async ({ authState, mutate }) => {
+  if (!authState) {
+    const token = await AsyncStorage.getItem(TOKEN_KEY, {});
+    const refreshToken = await AsyncStorage.getItem(REFRESH_TOKEN_KEY, {});
+    if (token && refreshToken) {
+      return { token, refreshToken };
+    }
+    return null;
+  }
+
+  return null;
+}
+```
+
+### Configuring `addAuthToOperation`
+
+The purpose of `addAuthToOperation` is to take apply your auth state to each request. Note that the format of the `authState` will be whatever
+you've returned from `getAuth` and not at all constrained by the exchange:
+
+```js
+const addAuthToOperation = ({
+  authState,
+  operation,
+}) => {
+  if (!authState || !authState.token) {
+    return operation;
+  }
+
+  const fetchOptions =
+    typeof operation.context.fetchOptions === 'function'
+      ? operation.context.fetchOptions()
+      : operation.context.fetchOptions || {};
+
+  return {
+    ...operation,
+    context: {
+      ...operation.context,
+      fetchOptions: {
+        ...fetchOptions,
+        headers: {
+          ...fetchOptions.headers,
+          "Authorization": authState.token,
+        },
+      },
+    },
+  };
+}
+```
+
+First we check that we have an `authState` and a `token`. Then we apply it to the request `fetchOptions` as an `Authorization` header.
+The header format can vary based on the API (e.g using `Bearer ${token}` instead of just `token`) which is why it'll be up to you to add the header
+in the expected format for your API.
+
+### Configuring `didAuthError`
+
+This function lets the exchange know what is defined to be an API error for your API:
+
+```js
+const didAuthError = ({ error }) => {
+  return error.graphQLErrors.some(
+    e => e.extensions?.code === 'FORBIDDEN',
+  );
+}
+```
+
+For most graphQL APIs, the auth error is communicated as an error code, however it may also be a 401 HTTP response:
+
+```js
+const didAuthError = ({ error }) => {
+  return error.graphQLErrors.some(
+    e => error.response.status === 401,
+  );
+},
+```
+
+Then `didAuthError` returns `true`, it will trigger the exchange to trigger the logic for asking for reauthentication via `getAuth`.
+
+### Configuring `getAuth` (after `didAuthError` returns `true`)
+
+If your API doesn't support any sort of token refresh, this is where you should simply log the user out.
+
+```js
+const getAuth = async ({ authState }) => {
+  if (!authState) {
+    const token = localStorage.getItem('token');
+    const refreshToken = localStorage.getItem('refreshToken');
+    if (token && refreshToken) {
+      return { token, refreshToken };
+    }
+    return null;
+  }
+
+  logout();
+
+  return null;
+}
+```
+
+Here, `logout()` should clear your persisted storage of any tokens and set the `isLoggedIn` property as described above to `false`.
+
+If you do have a token refresh endpoint, and a refresh token, you can attempt to get a new token for the user:
```

Comment:
```suggestion
Here, `logout()` is a placeholder that is called when we got an error, so that we can redirect to a login page again and clear our tokens from local storage or otherwise.

If we had a way to refresh our token using a refresh token, we can attempt to get a new token for the user first:
```

Just rewriting to keep the first-person plural consistent